File: databases/rocksdb/python/src/recipe_batch_write.py
"""
Recipe: WriteBatch Operations
Database: RocksDB
Description: Demonstrates using WriteBatch for high-throughput atomic writes,
    batching multiple puts and deletes in a single commit.

Usage: python src/recipe_batch_write.py
"""

import os
import shutil
import rocksdbpy


def main():
    db_path = "/tmp/rocksdb_recipe_batch_write"

    # 1. Setup — open the database
    db = rocksdbpy.open_default(db_path)

    # 2. WriteBatch — stage multiple operations atomically
    batch = rocksdbpy.WriteBatch()
    for i in range(1, 6):
        key = f"item:{i}".encode()
        value = f"value-{i}".encode()
        batch.add(key, value)
    batch.delete(b"item:3")  # Remove item:3 from the batch
    db.write(batch)

    # 3. Read back — verify batch results
    print("After WriteBatch:")
    for i in range(1, 6):
        key = f"item:{i}".encode()
        value = db.get(key)
        status = value.decode() if value else "(deleted)"
        print(f"  {key.decode()} => {status}")

    # 4. Cleanup
    db.close()
    shutil.rmtree(db_path, ignore_errors=True)
    print("\nDone.")


if __name__ == "__main__":
    main()

File: databases/rocksdb/rust/src/bin/recipe_batch_write.rs
//! Recipe: WriteBatch Operations
//! Database: RocksDB
//! Description: Demonstrates using WriteBatch for high-throughput atomic writes,
//!     batching multiple puts and deletes in a single commit.
//!
//! Usage: cargo run --bin recipe_batch_write

use rocksdb::{DB, Options, WriteBatch};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let db_path = "/tmp/rocksdb_recipe_batch_write";

    // 1. Setup — open the database
    {
        let db = DB::open_default(db_path)?;

        // 2. WriteBatch — stage multiple operations atomically
        let mut batch = WriteBatch::default();
        for i in 1..=5 {
            let key = format!("item:{}", i);
            let value = format!("value-{}", i);
            batch.put(key.as_bytes(), value.as_bytes());
        }
        batch.delete(b"item:3"); // Remove item:3 from the batch
        db.write(batch)?;

        // 3. Read back — verify batch results
        println!("After WriteBatch:");
        for i in 1..=5 {
            let key = format!("item:{}", i);
            match db.get(key.as_bytes())? {
                Some(value) => println!("  {} => {}", key, String::from_utf8(value)?),
                None => println!("  {} => (deleted)", key),
            }
        }
    }

    // 4. Cleanup
    let _ = DB::destroy(&Options::default(), db_path);
    println!("\nDone.");
    Ok(())
}
